src/olympia/blocklist/tasks.py

```diff
@@ -88,28 +97,62 @@ def monitor_remote_settings():
+    # and deserialize them here back to the enum values.
+    filter_list: List[BlockType] = (
+        [] if filter_list is None else [BlockType[filter] for filter in filter_list]
+    )
     bucket = settings.REMOTE_SETTINGS_WRITER_BUCKET
     server = RemoteSettings(
         bucket, REMOTE_SETTINGS_COLLECTION_MLBF, sign_off_needed=False
     )
     mlbf = MLBF.load_from_storage(generation_time, error_on_missing=True)
+    is_base = len(filter_list) > 0
+    oldest_base_filter_id: int | None = None
+
     if is_base:
-        # clear the collection for the base - we want to be the only filter
-        server.delete_all_records()
-        statsd.incr('blocklist.tasks.upload_filter.reset_collection')
-        # Then the bloomfilter
-        data = {
-            'key_format': MLBF.KEY_FORMAT,
-            'generation_time': generation_time,
-            'attachment_type': BLOCKLIST_RECORD_MLBF_BASE,
-        }
-        with mlbf.storage.open(mlbf.filter_path, 'rb') as filter_file:
-            attachment = ('filter.bin', filter_file, 'application/octet-stream')
-            server.publish_attachment(data, attachment)
-            statsd.incr('blocklist.tasks.upload_filter.upload_mlbf')
+        for block_type in filter_list:
+            data = {
+                'key_format': MLBF.KEY_FORMAT,
+                'generation_time': generation_time,
+                'attachment_type': BLOCKLIST_RECORD_MLBF_BASE(block_type),
+            }
+            with mlbf.storage.open(mlbf.filter_path(block_type), 'rb') as filter_file:
+                attachment = ('filter.bin', filter_file, 'application/octet-stream')
+                server.publish_attachment(data, attachment)
+            base_filter_id: int | None = get_config(
+                MLBF_BASE_ID_CONFIG_KEY(block_type), int_value=True
+            )
+            if base_filter_id is not None:
+                if oldest_base_filter_id is None:
+                    oldest_base_filter_id = base_filter_id
+                else:
+                    oldest_base_filter_id = min(oldest_base_filter_id, base_filter_id)
+
+        statsd.incr('blocklist.tasks.upload_filter.upload_mlbf')
         statsd.incr('blocklist.tasks.upload_filter.upload_mlbf.base')
-    else:
+
+        if oldest_base_filter_id is not None:
+            for record in server.records():
+                record_time = int(
+                    record['stash_time']
+                    if 'stash_time' in record
+                    else record['generation_time']
+                    if 'generation_time' in record
+                    # If we don't have a generation time to use
+                    # then setting the base filter id will ensure
+                    # the record is not deleted.
+                    else oldest_base_filter_id
+                )
+                if record_time < oldest_base_filter_id:
+                    server.delete_record(record['id'])
```

Comment:
Re-thinking about this - and I am not saying this isn't already what's happening - we should:

1. delete any filter that we are going to re-upload to keep always 1 filter (per block type)
2. delete everything when both filters are re-generated
3. delete stash records older than the oldest of the two filters

Reply:
- Added logic for 1. We check any pre-existing attachment that matches the `filter_list` and delete them.
- I did not include logic explicitly for 2 because it should happen anyway
- 3 was already implemented